Ensure desired state of world populator runs before volume reconstructor

wongma7 opened the pull request:
If the kubelet's volumemanager reconstructor for actual state of world runs before the desired state of world has been populated, the pods in the actual state of world will have some incorrect volume information: namely outerVolumeSpecName, which if incorrect leads to part of the issue here https://github.com/kubernetes/kubernetes/issues/43515, because WaitForVolumeAttachAndMount searches the actual state of world with the correct outerVolumeSpecName and won't find it so reports 'timeout waiting....', etc. forever for existing pods. The comments acknowledge that this is a known issue

The all sources ready check doesn't work because the sources being ready doesn't necessarily mean the desired state of world populator added pods from the sources. So instead let's put the all sources ready check in the *populator*, and when the sources are ready, it will be able to populate the desired state of world and make "HasAddedPods()" return true. THEN, the reconstructor may run.

@jingxu97 PTAL, you wrote all of the reconstruction stuff

```release-note
Fix erroneous FailedSync and FailedMount events being periodically and indefinitely posted on Pods after kubelet is restarted
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44781)
<!-- Reviewable:end -->

jingxu97 commented:
@wongma7 Thanks for fixing the issue. Your approach should work, I just has one small concern. If we put allSources.Ready check before loop though the pod list to populate desired state, it might delay the reconciler to start mount the volumes. In case of some error and allSources becomes ready after a long period, desiredStatePopulator will not process any pods and their volumes. 

I am thinking maybe we could put allSourcesReady check in the HasAddedPods function. This function will return true only if allSourceReady and populator finishes looping the pod list. In populatorLoopFunc, hasAddedPods is set to false at the begining, and change to true at the end of the function.

wongma7 commented:
@jingxu97 in my testing, checking allSourcesReady in HasAddedPods still doesn't guarantee that the desired state of world is ready before reconciler's reconstruction. The timing is quite tight, see this example:

I0424 11:48:09.384200   19213 volume_manager.go:241] The desired_state_of_world populator starts
I0424 11:48:14.399392   19213 reconciler.go:153] Desired state of world has been populated with pods, starting reconstruct state function
I0424 11:48:14.400094   19213 reconciler.go:483] Reconciler sync states: could not find pod information in desired or actual states or pending operation, update it in both states: &{volumeName:kubernetes.io/aws-ebs/aws://us-east-1d/vol-06425954fe553e31d podName:2d103d0c-2905-11e7-ae76-0e06794f72a4 volumeSpec:0xc421083280 outerVolumeSpecName:pvc-291faaf2-2905-11e7-ae76-0e06794f72a4 pod:0xc421ba7200 pluginIsAttachable:true volumeGidValue: devicePath: reportedInUse:false mounter:0xc421babbc0}
I0424 11:48:14.404771   19213 desired_state_of_world_populator.go:286] Added volume "dynamic" (volSpec="pvc-291faaf2-2905-11e7-ae76-0e06794f72a4") for pod "2d103d0c-2905-11e7-ae76-0e06794f72a4" to desired state.

It's like this:
1. Start kubelet
2. The populator runs multiple times, but probably doesn't find any pods because sources aren't ready. After x run, HasAddedPods returns true because sources are ready and the populator loop has run many times already. Now that sources are ready, the populator will surely find all pods in the *next* loop, 100 ms later, but chances are, up until now it hasn't found any pods.
3. Reconciler is running on the same 100 ms period. It now sees HasAddedPods return true and so runs. But DSW doesn't have all the pods, so some volumes in ASW get wrong information.
4. Populator runs again and finds pods because sources are ready. This is only <100 ms too late, but still too late, ASW has the wrong information.

I think to avoid a long delay, we could add a 15 seconds timeout where the populator must run if allSourcesReady has not happened by then. And then the reconciler reconstruction should only run after the populator has run once when sources were ready. Let me update the patch...

wongma7 commented:
@jingxu97 patch updated, PTAL.

I made it so desired state of world populator is not blocked by sources.AllReady as recommended. I didn't change the HasAddedPods function because as I said in my testing that didn't work out, instead I changed how 'dswp.hasAddedPods' gets set: do it right after the populator has run with all ready sources.

jingxu97 commented:
Just a small comment, otherwise lgtm.
Thanks for the fix!

gnufied commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -100,24 +110,33 @@ type desiredStateOfWorldPopulator struct {
+		if sourcesReady.AllReady() {
+			defer func() { dswp.hasAddedPods = true }()
+		}
+
```
Are we using `defer` unnecessarily here? How about adding:

```
if sourcesReady.AllReady() {
			dswp.hasAddedPods = true
		}
```

line after `findAndAddNewPods` invocation?

wongma7 replied:
It's not the same, because we have to make sure to set dswp.hasAddedPods = true only if sources were ready at the time `findAndAddNewPods` was called. Otherwise it's possible that sources weren't ready, findAndAddNewPods gets called and finds nothing, and then dswp.hasAddedPods is set to true just one loop too soon.

wongma7 replied:
whoops, my above comment is addresisng the case where we do 
```
if sourcesReady.AllReady() {
	dswp.hasAddedPods = true
}
```
*after* findAndAddNewPods.

if we do it *before*, there is the possibility that ASW reconciler will run too early, when the populator is still in the middle of adding pods from all the sources.

wongma7 commented:
@jingxu97 I don't see your comment, could you try submitting it again? Thanks : )

wongma7 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -80,6 +80,8 @@ type Reconciler interface {
 // desiredStateOfWorld - cache containing the desired state of the world
 // actualStateOfWorld - cache containing the actual state of the world
+// populatorHasAddedPods - checker for whether the populator has added pods to
+//   the desiredStateOfWorld yet
```
I updated this comment.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -80,6 +80,9 @@ type Reconciler interface {
 // desiredStateOfWorld - cache containing the desired state of the world
 // actualStateOfWorld - cache containing the actual state of the world
+// populatorHasAddedPods - checker for whether the populator has finished
+//   adding pods to the desiredStateOfWorld cache at least once after sources
```
remove the extra empty space?

wongma7 replied:
I'm copying the other multi-line parameter comments which are indented by 2 spaces if you mean the spaces at the beginning

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -94,6 +97,7 @@ func NewReconciler(
 	nodeName types.NodeName,
 	desiredStateOfWorld cache.DesiredStateOfWorld,
 	actualStateOfWorld cache.ActualStateOfWorld,
+	populatorHasAddedPods func() bool,
```
just want to double check what is the benefit of using a function instead of passing the pointer to desiredStatePopulator?

wongma7 replied:
Imo it's cleaner to expose it via the interface.

And actually this made me think, the bool is not threadsafe, we could read and write to the bool at the same time? I need to add a lock I think

gnufied replied:
I don't think we need lock before reading/writing that variable. afaict - it is not being called concurrently by any other goroutine. 

The volume manager calls `dswp.Run` and that in turn loops the function infinitely between the sleep durations. I don't see any concurrent execution of `dswp.Run` ..

wongma7 replied:
OK, I added a lock which justifies using a function. It's quite ugly to have to lock, unlock, set and check it forever, when we only need to do it once but yeah, this should prevent any concurrency problem between populator & reconciler at startup

wongma7 replied:
@gnufied I highly highly doubt it is ever going to happen ,but I don't think it's impossible for dswp.Run's write and reconciler.Run's read to coincide. :/ You're right it would have to happen the first time at startup, after that there's no chance of concurrent read/write.

gnufied replied:
Ah, I missed that reconciler is reading the value. In that case yeah, we perhaps do need the lock.

euank commented:
/unassign

Commit:
Ensure desired state of world populator runs before volume reconstructor

wongma7 commented:
I've updated the PTAL.

With the addition of the lock, things got a tad more complicated so I reworked the patch to be absolutely clear about what it's doing. The populator needs to set hasAddedPods only once, then it's done touching that variable and can resume normal operation. Similarly, the reconciler needs to check hasAddedPods only once, then it's done and can resume.

wongma7 commented:
@jingxu97 ping, PTAL when you get a chance :) The spirit of the patch is the same, I just tried to make it clearer & safer prompted by the comments. @saad-ali can you please /approve this PR, you are the only approver in this directory and for some reason the bot decided to pick on somebody from a parent dir.

I would like to get this cherry-picked into 1.6.3 once it's good to go, it is trivial to reproduce and leads to really bad UX IMO

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -132,20 +138,26 @@ type reconciler struct {
+		rc.reconciliationLoopFunc(sourcesReady, rc.populatorHasAddedPods())()
+		return rc.populatorHasAddedPods(), nil
+	}, stopCh)
+	wait.Until(rc.reconciliationLoopFunc(sourcesReady, true), rc.loopSleepDuration, stopCh)
```
Is the purpose of this change is to avoid keep checking populatorHasAddedPods? Is this very necessary?

wongma7 replied:
yes. No it's not technically necessary, I thought it would make the code clearer where/why populatroHasAddedPods is getting called since after discussion with gnufied it was not immediately clear. I can trim it back if you want, I could see how this looks like premature optimization :)

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -132,20 +138,26 @@ type reconciler struct {
+		// reconstruct process may add incomplete volume information and cause confusion. In addition, if the
+		// desired state of world has not been populated yet, the reconstruct process may clean up pods' volumes
+		// that are still in use because desired state of world does not contain a complete list of pods.
+		if sourcesReady.AllReady() && populatorHasAddedPods && time.Since(rc.timeOfLastSync) > rc.syncDuration {
```
I think populatorHasAddedPods returns true indicates sourcesReady.AllReady()?

wongma7 replied:
yes, so I have removed the sourcesReady arg from reconciler.Run

jingxu97 commented:
/approve

jingxu97 commented:
sorry for the delay.
/lgtm

jingxu97 commented:
@saad-ali could you approve this? Thanks!

wongma7 commented:
@jingxu97 sorry, forgot about the integration test reconciler_test.go. Had to make reconcilerLoopSleepDuration nonzero

saad-ali commented:
/lgtm
/approve

wongma7 commented:
please re-add lgtm, (bazel) tests should pass now, thank you!

jingxu97 commented:
/lgtm

wongma7 commented:
@k8s-bot bazel test this
W0510 00:49:48.944] fatal: Unable to create '/root/.cache/git/k8s.io/kubernetes/index.lock': File exists.

gnufied commented:
@k8s-bot bazel test this

gnufied commented:
@k8s-bot bazel test this

jingxu97 commented:
@wongma7  could you cherrypick the PR to 1.5 and 1.6?

wongma7 commented:
Ok, on it. will ping you in the PRs.

On Thu, May 11, 2017 at 2:39 PM, Jing Xu <notifications@github.com> wrote:

> @wongma7 <https://github.com/wongma7> could you cherrypick the PR to 1.5
> and 1.6?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44781#issuecomment-300879935>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMgP-JFhFVgpPdhTEyFEYs_PukE5BBweks5r41XzgaJpZM4NExDf>
> .
>

wongma7 commented:
/release-note

DreadPirateShawn commented:
Am I reading correctly that the cherry-pick back into Kube 1.6 failed?

So should I expect to see this in Kube 1.7 but no earlier?

wongma7 commented:
It was cherry picked to 1.5 and 1.6 both, not sure the exact .z patch number

DreadPirateShawn commented:
Oh? The "Merge pull request" for 1.6 has a red X with hover text "Failure: k8s-merge-robot needs CLA" -- was that a transient issue, and Github simply doesn't update the status to remove the failure text, then?

DreadPirateShawn commented:
Update -- I see the changelog citing this PR here -- https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.6.md#v165

Fair enough. Filing a fresh ticket shortly...
